code/game/machinery/firealarm.dm

```diff
@@ -326,3 +326,109 @@
 	if (!party_overlay)
 		party_overlay = iconstate2appearance('icons/turf/areas.dmi', "party")
 	A.add_overlay(party_overlay)
+
+/obj/machinery/sprinkler
+	name = "foam dispenser"
+	desc = "An automated sprinkler cabable of detecting fire and spray coolant."
+	icon = 'icons/obj/device.dmi'
+	icon_state = "sprinkler0"
+	layer = PROJECTILE_HIT_THRESHHOLD_LAYER
+	plane = FLOOR_PLANE
+	max_integrity = 100
+	armor = list("melee" = 50, "bullet" = 20, "laser" = 20, "energy" = 20, "bomb" = 0, "bio" = 0, "rad" = 0, "fire" = 80, "acid" = 30)
+	var/obj/item/reagent_containers/glass/beaker/internal
+	var/refill_loaded = FALSE
+	req_access = list(ACCESS_ATMOSPHERICS)
+	var/cooldown = 0
+
+/obj/machinery/sprinkler/Initialize()
+	. = ..()
+	internal = new /obj/item/reagent_containers/glass/beaker(src)
+
+/obj/machinery/sprinkler/interact(mob/user)
+	if(!allowed(user))
+		to_chat(user, "<span class='danger'>Access denied.</span>")
+		return
+	dispense()
+
+/obj/machinery/sprinkler/fire_act()
+	.=..()
+	if(refill_loaded)
+		dispense()
+
+/obj/machinery/sprinkler/proc/dispense()
+	if(cooldown < world.time && refill_loaded)
+		var/datum/effect_system/smoke_spread/chem/smoke_machine/smoke = new()
+		smoke.set_up(internal.reagents, 5, 8,  get_turf(src))
+		smoke.start()
+		cooldown = world.time + 200
+		if(!internal.reagents)
+			refill_loaded = FALSE
+			update_icon()
+
+/obj/machinery/sprinkler/update_icon()
+	. = ..()
+	if(refill_loaded)
+		icon_state = "sprinkler1"
+		return
+	icon_state = "sprinkler0"
+
+/obj/item/sprinkler_refill

```

Comment:
Rather than having a dedicated sprinkler refill device, if you switch to the sprinkler itself having a `reagents`, you can make it refillable with theoretically _any_ chemical, which has potential for hilarity.